cranelift/codegen/src/isa/zkasm/inst/emit.rs

```diff
@@ -455,6 +455,13 @@ impl MachInstEmit for Inst {
                     sink,
                 );
             }
+            Inst::Ineg { rd, rs1 } => {
+                let rs = allocs.next(*rs1);
+                let rd = allocs.next(rd.to_reg());
+                // TODO(nagisa): this wants to conditionally use the `SUB` instruction instead (if
+                // the negation is for 64 bits)
```

Comment:
Same as in another PR inline substraction works.

Reply:
Does it also work with register operands? E.g. `0n - A => A`?

Reply:
Looks like it does not:

```
  42n => A              # A = 42
  4294967296n - A => A  # A = 0x1_0000_0000 - 42
  4294967254n => B      # B = 0xffff_ffd6
  B :ASSERT
```

fails with

```
Assert does not match [w:9 zkPC:10 sub_test.zkasm:13] (op:4294967254 A:18446744073709551575)
```

meaning that the subtraction happened seems to have been done in 32-bits (0 - 42) and then sign extended to 64-bits when writing the result, rather than being done in 64-bits in the first place. It works correctly when both operands are literal though.

Reply:
Actually, I’m not actually sure about the sign extension, as e.g. `0n - A` produces `0xfffffffeffffffd7`. it looks like it may be taking a carry from the upper half, but not actually computing with it considered properly? Might be a bug, might be not, I don’t know.